File: CafeApp/app/src/main/java/by/ivan/CafeApp/ui/data/repo/TableRepository.kt
package by.ivan.CafeApp.ui.data.repo

import android.content.Context
import by.ivan.CafeApp.ui.data.datasource.TableLocalDatasource
import by.ivan.CafeApp.ui.data.datasource.TableRemoteDatasource
import by.ivan.CafeApp.ui.data.local.entity.TableLocalModel
import by.ivan.CafeApp.ui.data.local.entity.TableVersionLocalModel
import by.ivan.CafeApp.ui.data.models.State
import by.ivan.CafeApp.ui.data.models.Table
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class TableRepository @Inject constructor(
    private val tableRemoteDatasource: TableRemoteDatasource,
    private val tableLocalDatasource: TableLocalDatasource,
    private val tableVersionsRepository: TableVersionsRepository,
    private val connectionStateRepository: ConnectionStateRepository,
    @ApplicationContext private val context: Context
) {
    suspend fun getTables(): List<Table> {
        if(connectionStateRepository.checkForInternet(context)){
            var remoteVersion = tableVersionsRepository.getRemoteTableVersion("Table")
            var localVersion = tableVersionsRepository.getLocalTableVersion("Table")

            if (localVersion.version == -1) {
                val tablesRemote = tableRemoteDatasource.getTables().items
                val tableLocalModel = tablesRemote.map {
                    TableLocalModel(
                        id = it.id,
                        title = it.title,
                        state = State.values().first { vals ->  vals.i== it.state}
                    )
                }

                tableLocalDatasource.removeAll()
                tableLocalDatasource.saveTables(tableLocalModel)

                remoteVersion.apply {
                    tableVersionsRepository.insert(
                        TableVersionLocalModel
                            (id = this.id, tableName = this.tableName, version = this.version)
                    )
                }
            } else if ((remoteVersion.version != -1 && remoteVersion.version != localVersion.version)) {
                val tablesRemote = tableRemoteDatasource.getTables().items
                val tableLocalModel = tablesRemote.map {
                    TableLocalModel(
                        id = it.id,
                        title = it.title,
                        state = State.valueOf(it.state.toString())
                    )
                }

                tableLocalDatasource.removeAll()
                tableLocalDatasource.saveTables(tableLocalModel)

                localVersion.version = remoteVersion.version
                tableVersionsRepository.edit(localVersion)
            }


            val tables = tableLocalDatasource.getTables().map {
                Table(
                    id = it.id,
                    title = it.title,
                    state = 0
                )
            }

            return tables
        }
        else{
            val tables = tableLocalDatasource.getTables().map {
                Table(
                    id = it.id,
                    title = it.title,
                    state = 0
                )
            }

            return tables
        }
    }
}

File: Cafe.Service/Cafe.Service/Models/Table/TableRepository.cs
using Cafe.Service.Models.Shared;
using Microsoft.EntityFrameworkCore;

namespace Cafe.Service.Models.Table
{
    public class TableRepository
    {
        private ApplicationContext _context;

        public TableRepository(ApplicationContext context)
        {
            _context = context;
        }

        public DbSet<Table> Get()
        {
            return _context.Tables;
        }

        public Table Add(Table table)
        {
            _context.Tables.Add(table);
            _context.SaveChanges();
            return table;
        }

        public Table Edit(Table table)
        {
            var item = _context.Tables.FirstOrDefault(c => c.Id == table.Id);
            if (item == null)
            {
                throw new ArgumentOutOfRangeException();
            }
            item.Title = table.Title;
            item.State = table.State;
            _context.SaveChanges();
            return item;
        }

        public Table Remove(Table table)
        {
            _context.Tables.Remove(table);
            _context.SaveChanges();
            return table;
        }
    }
}
